source/fuzz/transformation_wrap_vector_synonym.cpp

```diff
@@ -0,0 +1,172 @@
+// Copyright (c) 2021 Shiyu Liu
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "source/fuzz/transformation_wrap_vector_synonym.h"
+#include <algorithm>
+#include "source/fuzz/data_descriptor.h"
+#include "source/fuzz/fuzzer_util.h"
+#include "source/fuzz/instruction_descriptor.h"
+#include "source/opt/instruction.h"
+
+namespace spvtools {
+namespace fuzz {
+
+TransformationWrapVectorSynonym::TransformationWrapVectorSynonym(
+    protobufs::TransformationWrapVectorSynonym message)
+    : message_(std::move(message)) {}
+
+TransformationWrapVectorSynonym::TransformationWrapVectorSynonym(
+    uint32_t instruction_id, uint32_t vector_operand1, uint32_t vector_operand2,
+    uint32_t vec_id, uint32_t pos) {
+  message_.set_instruction_id(instruction_id);
+  message_.set_vector_operand1(vector_operand1);
+  message_.set_vector_operand2(vector_operand2);
+  message_.set_vec_id(vec_id);
+  message_.set_scalar_position(pos);
+}
+
+bool TransformationWrapVectorSynonym::IsApplicable(
+    opt::IRContext* ir_context,
+    const TransformationContext& transformation_context) const {
+  auto instruction =
+      ir_context->get_def_use_mgr()->GetDef(message_.instruction_id());
+  auto vec1 = ir_context->get_def_use_mgr()->GetDef(message_.vector_operand1());
+  auto vec2 = ir_context->get_def_use_mgr()->GetDef(message_.vector_operand2());
+
+  // |instruction_id| must refer to an existing instruction.
+  if (instruction == nullptr) {
+    return false;
+  }
+
+  std::unordered_set<SpvOp> valid_opcodes{SpvOpIAdd, SpvOpISub, SpvOpIMul,
+                                          SpvOpSDiv, SpvOpUDiv, SpvOpFAdd,
+                                          SpvOpFSub, SpvOpFMul, SpvOpFDiv};
+
+  if (!valid_opcodes.count(instruction->opcode())) {
+    return false;
+  }
+
+  // |vector_operand1| and |vector_operand2| must exist.
+  if (vec1 == nullptr || vec2 == nullptr) {
+    return false;
+  }
+
+  // The 2 vectors must be the same valid vector type.
+  auto vec1_type_id = vec1->type_id();
+  auto vec2_type_id = vec2->type_id();
+  auto vec_type = ir_context->get_type_mgr()->GetType(vec1_type_id);
```

Comment:
The type manager library isn't that nice. Although we do sometimes use it in the code base we have recently been trying not to.

Instead, can you please get the type instruction - i.e., look up vec1_type_id in the def use manager?

Reply:
Yes, I do have the following combo when I wrote it at first: 
```
  auto vec_type = ir_context->get_def_use_mgr()->GetDef(vec1_type_id)->opcode();

  if (vec_type != SpvOpTypeVector) {
    return false;
  }

```
Later I changed it to the one we have now. I know they practically are the same, but is the difference mainly coming from a readability stand point or there are some performance differences as well? Because for me the opcode one is actually easier to understand, and it avoids trying to cast different type into vector if the type is not a vector.